gitlab-runner/src/run.rs

```diff
@@ -54,36 +55,74 @@ where
         Ok(())
     };
 
-    let upload = !cancel_token.is_cancelled()
-        && response.artifacts.get(0).map_or(false, |a| match a.when {
-            ArtifactWhen::Always => true,
-            ArtifactWhen::OnSuccess => script_result.is_ok(),
-            ArtifactWhen::OnFailure => script_result.is_err(),
-        });
-
-    let r = if upload {
-        if let Ok(mut uploader) = Uploader::new(client, &build_dir, response) {
-            let r = handler.upload_artifacts(&mut uploader).await;
-            if r.is_ok() {
-                uploader.upload().await.and(script_result)
-            } else {
-                r
+    let mut overall_result = script_result;
+
+    for artifact in response.artifacts.iter().take(1) {
+        let upload = !cancel_token.is_cancelled()
+            && match artifact.when {
+                ArtifactWhen::Always => true,
+                ArtifactWhen::OnSuccess => script_result.is_ok(),
+                ArtifactWhen::OnFailure => script_result.is_err(),
+            };
+        if !upload {
+            continue;
+        }
+        if let Ok(mut uploader) = Uploader::new(client.clone(), &build_dir, response.clone()) {
+            let mut available_files: Vec<U> = handler.get_uploadable_files().await?;
+            let mut res = Ok(());
+            let mut uploaded = 0;
+
+            for path in artifact.paths.iter() {
+                let matches = match glob::Pattern::new(path) {
+                    Ok(pattern) => available_files
+                        .iter()
+                        .enumerate()
+                        .filter(|(_, f)| pattern.matches(&f.get_path()))
+                        .map(|(i, _)| i)
+                        .rev()
```

Comment:
It's a little evil, this really would be much nicer with `drain_filter` if it was stabilized. But this allows us to iterate over the indices, removing them from `available_files`, and not invalidate the other indices: since they're in reverse order, processing tail first keeps all other indices correct until we remove them.

Reply:
uhggg; yeah please try to avoid that; would be better to get an iterator over matching files and upload those rather then fool around with indexes whichis way too subtle